invenio_files_rest/storage.py

```diff
@@ -127,3 +132,83 @@ def send_file(self, restricted=False):
                 content_md5=self.file.checksum)
         except Exception as e:
             raise StorageError('Could not send file: {}'.format(e))
+
+
+class AmazonS3(Storage):
+    def __init__(self, object, fileinstance):
+        super(AmazonS3, self).__init__(object, fileinstance)
+        if "AWS_KEY" in current_app.config:
+            raise StorageError("Missing AWS Key")
+        if "AWS_SECRET" in current_app.config:
+            raise StorageError("Missing AWS Secret")
+        self.aws_connection = S3Connection(current_app.config.get("AWS_KEY", "AWS_KEY"),
+                                           current_app.config.get("AWS_SECRET", "AWS_SECRET"))
+
+    def get_size(self):
+        bucket_name, key_name = self.obj.file.uri.replace("s3:", "").split(":")
+        bucket = self.aws_connection.get_bucket(bucket_name)
+        return bucket.lookup(key_name).size
+
+    def open(self):
+
+        temp = tempfile.NamedTemporaryFile()
+        try:
+            loc = self.obj.bucket.location.uri
+            bucket = self.aws_connection.get_bucket(loc)
+            element = Key(bucket)
+            key_name = self.obj.file.uri.replace("s3:", "").split(":")[1]
+            element.key = key_name
+            element.get_contents_to_file(temp)
+            temp.seek(0)
+            return temp
+        except:
+            temp.close()
+            return None
+
+    def send_file(self, filename):
+        """Send the file to the client.
+
+        This returns a flask response that will eventually result in
+        the user being offered to download the file (or view it in the
+        browser).  Depending on the storage backend it may actually
+        send a redirect to an external URL where the file is available.
+
+        :param filename: The file name to use when sending the file to
+                         the client.
+        """
+        loc = self.obj.bucket.location.uri
+        bucket = self.aws_connection.get_bucket(loc)
+        element = Key(bucket)
+        element.key = self.obj.file.uri
+        return redirect(element.generate_url(expires_in=300))
+
+    def save(self, incoming_stream, size=None, chunk_size=None):
+
+        # Retrieve the bucket name
+        loc = self.obj.bucket.location.uri
+        bucket = self.aws_connection.get_bucket(loc)
+        element = Key(bucket)
+        element.key = "".join((incoming_stream.name, self.make_path()))
+
+        chunk_size = chunk_size or 1024 * 64
+        m = hashlib.md5()
+        bytes_written = 0
+
+        while 1:
```

Comment:
If you wrap ``element`` in something with a ``write`` method you can use ``_save_stream`` directly.